File: services/ingestion-openmeteo/app/Domain/Shared/Event/WeatherSnapshotFetched.php
<?php

namespace App\Domain\Shared\Event;

use App\Application\Weather\DTO\WeatherSnapshotData;
use App\Domain\Weather\Location;
use Ramsey\Uuid\Uuid;

final readonly class WeatherSnapshotFetched implements DomainEvent
{
    private string $id;
    private string $occurredAtValue;

    public function __construct(
        private Location $location,
        private WeatherSnapshotData $snapshotData,
    ) {
        $this->id = Uuid::uuid7()->toString();
        $this->occurredAtValue = now()->toIso8601String();
    }

    /**
     * @inheritDoc
     */
    public function eventId(): string
    {
        return $this->id;
    }

    /**
     * @inheritDoc
     */
    public function eventName(): string
    {
        return 'weather.snapshot.fetched';
    }

    /**
     * @inheritDoc
     */
    public function eventVersion(): int
    {
        return 1;
    }

    /**
     * @inheritDoc
     */
    public function occurredAt(): string
    {
        return $this->occurredAtValue;
    }

    /**
     * @inheritDoc
     */
    public function payload(): array
    {
        return $this->snapshotData->toArray();
    }

    /**
     * Lokacija kojoj event pripada.
     * @return Location
     */
    public function location(): Location
    {
        return $this->location;
    }
}

File: services/ingestion-weatherapi/internal/domain/weather/weather_snapshot_fetched.go
package weather

import (
	"time"

	"github.com/google/uuid"
)

/*
WeatherSnapshotFetched predstavlja domain event koji nastaje kada servis
uspešno preuzme i normalizuje vremenski snapshot za datu lokaciju.
*/
type WeatherSnapshotFetched struct {
	id         string
	occurredAt string
	location   Location
	snapshot   WeatherSnapshotData
}

/*
NewWeatherSnapshotFetched pravi novi domain event za uspešno preuzet snapshot.
*/
func NewWeatherSnapshotFetched(
	location Location,
	snapshot WeatherSnapshotData,
) *WeatherSnapshotFetched {
	return &WeatherSnapshotFetched{
		id:         uuid.NewString(),
		occurredAt: time.Now().UTC().Format(time.RFC3339),
		location:   location,
		snapshot:   snapshot,
	}
}

/*
EventID vraća jedinstveni identifikator eventa.
*/
func (e *WeatherSnapshotFetched) EventID() string {
	return e.id
}

/*
EventName vraća canonical naziv domain eventa.
*/
func (e *WeatherSnapshotFetched) EventName() string {
	return "weather.snapshot.fetched"
}

/*
EventVersion vraća verziju event contract-a.
*/
func (e *WeatherSnapshotFetched) EventVersion() int {
	return 1
}

/*
OccurredAt vraća vreme nastanka eventa.
*/
func (e *WeatherSnapshotFetched) OccurredAt() string {
	return e.occurredAt
}

/*
Payload vraća canonical payload domen događaja.
*/
func (e *WeatherSnapshotFetched) Payload() map[string]any {
	return e.snapshot.ToMap()
}

/*
Location vraća lokaciju kojoj događaj pripada.
*/
func (e *WeatherSnapshotFetched) Location() Location {
	return e.location
}
